// @oceanprotocol/squid/dist/node/keeper/Keeper.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var DIDRegistry_1 = require("./contracts/DIDRegistry");
var Dispenser_1 = require("./contracts/Dispenser");
var Token_1 = require("./contracts/Token");
var conditions_1 = require("./contracts/conditions");
var templates_1 = require("./contracts/templates");
var managers_1 = require("./contracts/managers");
var EventHandler_1 = require("./EventHandler");
var Instantiable_abstract_1 = require("../Instantiable.abstract");
var Keeper = (function (_super) {
    __extends(Keeper, _super);
    function Keeper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Keeper.getInstance = function (config) {
        return __awaiter(this, void 0, void 0, function () {
            var keeper, _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
            return __generator(this, function (_l) {
                switch (_l.label) {
                    case 0:
                        keeper = new Keeper();
                        keeper.setInstanceConfig(config);
                        config.ocean.keeper = keeper;
                        _a = keeper;
                        return [4, Dispenser_1.default.getInstance(config)];
                    case 1:
                        _a.dispenser = _l.sent();
                        _b = keeper;
                        return [4, Token_1.default.getInstance(config)];
                    case 2:
                        _b.token = _l.sent();
                        _c = keeper;
                        return [4, DIDRegistry_1.default.getInstance(config)];
                    case 3:
                        _c.didRegistry = _l.sent();
                        _d = keeper;
                        return [4, managers_1.TemplateStoreManager.getInstance(config)];
                    case 4:
                        _d.templateStoreManager = _l.sent();
                        _e = keeper;
                        return [4, managers_1.AgreementStoreManager.getInstance(config)];
                    case 5:
                        _e.agreementStoreManager = _l.sent();
                        _f = keeper;
                        return [4, managers_1.ConditionStoreManager.getInstance(config)];
                    case 6:
                        _f.conditionStoreManager = _l.sent();
                        _g = keeper;
                        _h = {};
                        return [4, conditions_1.LockRewardCondition.getInstance(config)];
                    case 7:
                        _h.lockRewardCondition = _l.sent();
                        return [4, conditions_1.EscrowReward.getInstance(config)];
                    case 8:
                        _h.escrowReward = _l.sent();
                        return [4, conditions_1.AccessSecretStoreCondition.getInstance(config)];
                    case 9:
                        _g.conditions = (_h.accessSecretStoreCondition = _l.sent(),
                            _h);
                        _j = keeper;
                        _k = {};
                        return [4, templates_1.EscrowAccessSecretStoreTemplate.getInstance(config)];
                    case 10:
                        _j.templates = (_k.escrowAccessSecretStoreTemplate = _l.sent(),
                            _k);
                        keeper.utils = {
                            eventHandler: new EventHandler_1.EventHandler(config),
                        };
                        return [2, keeper];
                }
            });
        });
    };
    Keeper.prototype.getConditionByAddress = function (address) {
        return Object.values(this.conditions)
            .find(function (condition) { return condition.getAddress() === address; });
    };
    Keeper.prototype.getTemplateByName = function (name) {
        return Object.values(this.templates)
            .find(function (template) { return template.contractName === name; });
    };
    Keeper.prototype.getNetworkId = function () {
        return this.web3.eth.net.getId();
    };
    Keeper.prototype.getNetworkName = function () {
        return this.web3.eth.net.getId()
            .then(function (networkId) {
            var network = "Unknown";
            switch (networkId) {
                case 1:
                    network = "Main";
                    break;
                case 2:
                    network = "Morden";
                    break;
                case 3:
                    network = "Ropsten";
                    break;
                case 4:
                    network = "Rinkeby";
                    break;
                case 77:
                    network = "POA_Sokol";
                    break;
                case 99:
                    network = "POA_Core";
                    break;
                case 42:
                    network = "Kovan";
                    break;
                case 8996:
                    network = "Spree";
                    break;
                case 8995:
                    network = "Nile";
                    break;
                default:
                    network = "Development";
            }
            return network;
        });
    };
    return Keeper;
}(Instantiable_abstract_1.Instantiable));
exports.Keeper = Keeper;
exports.default = Keeper;
