// @oceanprotocol/squid/dist/node/squid.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var Config_1 = require("./models/Config");
exports.Config = Config_1.default;
var Account_1 = require("./ocean/Account");
exports.Account = Account_1.default;
var DID_1 = require("./ocean/DID");
exports.DID = DID_1.default;
var Ocean_1 = require("./ocean/Ocean");
exports.Ocean = Ocean_1.Ocean;
var Logger_1 = require("./utils/Logger");
exports.Logger = Logger_1.LoggerInstance;
var WebServiceConnectorProvider_1 = require("./utils/WebServiceConnectorProvider");
exports.WebServiceConnectorProvider = WebServiceConnectorProvider_1.default;
var Keeper_1 = require("./keeper/Keeper");
exports.Keeper = Keeper_1.default;
var templates = require("./keeper/contracts/templates");
exports.templates = templates;
var conditions = require("./keeper/contracts/conditions");
exports.conditions = conditions;
var utils = require("./utils");
exports.utils = utils;
__export(require("./ddo/DDO"));
var templates_1 = require("./keeper/contracts/templates");
exports.AgreementTemplate = templates_1.AgreementTemplate;
var conditions_1 = require("./keeper/contracts/conditions");
exports.Condition = conditions_1.Condition;
